### ISSUE TITLE
Inconsisten channels during document saving.

### ISSUE THREAD

**Issue by ont** [couchbase/sync_gateway#2995]
### Sync Gateway version
`"version":"Couchbase Sync Gateway/1.4.1(3;ddffd85)"`

### Operating system
Official docker container `1.4.1-3-comunity`

### Log output
https://gist.github.com/ont/844348c7bca6a1eb519ea5f3d3efb007

What happens:

1. push and pull replication receives 401
2. mobile app go to custom authentication server
3. authentication server creates user `13266487-63e4-4108-9199-76c1baee3d12` and creates sessions for push and pull replicators
4. mobile pushes new document `56ce883e-c3d6-419c-b374-6dc76cca17e9` which adds new channels to user (channel name equals to user's id `13266487-63e4-4108-9199-76c1baee3d12`)
5. sync-gateway tries to update user's profile and throws error (probably due to bad responses from couchbase server): 
```
2017-10-27T07:24:22.052Z WARNING: Error invalidating channels for user 13266487-63e4-4108-9199-76c1baee3d12: MCResponse status=TMPFAIL, opcode=SET, opaque=0, msg: Temporary failure -- db.(*Database).invalUserChannels() at database.go:964
```
6. no new channel was added to user
7. after that no error messages was sent to mobile client so it tries to upload all other documents
8. this documents needs user to have access to channel `13266487-63e4-4108-9199-76c1baee3d12` so we can see 403 errors in logs

### Expected behavior
Sync gateway tries to resend queries to couchbase and reject document saving if there are any problems (consistent database state). 

### Actual behavior
Now we have document `56ce883e-c3d6-419c-b374-6dc76cca17e9` which was uploaded to couchbase server, wrong channels for user `13266487-63e4-4108-9199-76c1baee3d12` and couple of rejected documents. Rejected documents **will not be reuploaded after channel fixation for user**. Channel fixation was made by uploading another document which gives channels access to the user. 

This problem was found during our pre-production tests and we do not have any good workaround.

### Some questions
1. It seems that this problem was fixed by 9c2f4a187420d228848eb1ebbd5e074541ae7863 . Will it be backported to `1.4.1` version?
2. When new release with this commit will be released?
3. Does new version `1.5.x` is backward-compatible with `1.4.1` version?

**Comment by adamcfraser** [comment]
I agree that the retry on the Set operation will prevent this issue in most situations (https://github.com/couchbase/sync_gateway/commit/9c2f4a187420d228848eb1ebbd5e074541ae7863). 

I don't anticipate we'll backport this to 1.4.1 - in 1.5 we've switched our underlying server client (from go-couchbase to gocb), and the fix has only been applied to gocb.

https://github.com/couchbase/sync_gateway/commit/9c2f4a187420d228848eb1ebbd5e074541ae7863 will be included in release 1.5.1, which is currently targeted for release in early 2018.  1.5.0 and 1.5.1 are both backward-compatible with 1.4.1 data.

**Comment by djpongh** [comment]
Fix will be in 1.5.1
